FrameMetadata/FrameMetadata/ViewController.swift

```diff
@@ -0,0 +1,231 @@
+    
+    func session(_ session: OTSession, streamCreated stream: OTStream) {
+        print("Session streamCreated: \(stream.streamId)")
+        if subscriber == nil {
+            doSubscribe(stream)
+        }
+    }
+    
+    func session(_ session: OTSession, streamDestroyed stream: OTStream) {
+        print("Session streamDestroyed: \(stream.streamId)")
+        if let subStream = subscriber?.stream, subStream.streamId == stream.streamId {
+            cleanupSubscriber()
+        }
+    }
+    
+    func session(_ session: OTSession, didFailWithError error: OTError) {
+        print("session Failed to connect: \(error.localizedDescription)")
+    }
+    
+}
+
+// MARK: - OTPublisher delegate callbacks
+extension ViewController: OTPublisherDelegate {
+    func publisher(_ publisher: OTPublisherKit, streamCreated stream: OTStream) {
+        print("Publishing")
+    }
+    
+    func publisher(_ publisher: OTPublisherKit, streamDestroyed stream: OTStream) {
+        if let subStream = subscriber?.stream, subStream.streamId == stream.streamId {
+            cleanupSubscriber()
+        }
+    }
+    
+    func publisher(_ publisher: OTPublisherKit, didFailWithError error: OTError) {
+        print("Publisher failed: \(error.localizedDescription)")
+    }
+    
+}
+
+// MARK: - OTSubscriber delegate callbacks
+extension ViewController: OTSubscriberDelegate {
+    func subscriberDidConnect(toStream subscriberKit: OTSubscriberKit) {
+        subscriber?.view?.frame = CGRect(x: 0, y: view.frame.width / kWidgetRatio, width: view.frame.width, height: view.frame.width / kWidgetRatio)
+        if let subsView = subscriber?.view {
+            view.addSubview(subsView)
+        }
+    }
+    
+    func subscriber(_ subscriber: OTSubscriberKit, didFailWithError error: OTError) {
+        print("Subscriber failed: \(error.localizedDescription)")
+    }
+    
+    func subscriberVideoDataReceived(_ subscriber: OTSubscriber) {
+    }
+}
+
+extension ViewController: ExampleVideoRenderDelegate {
+    func renderer(_ renderer: ExampleVideoRender, didReceiveFrame videoFrame: OTVideoFrame) {
+        guard let metadata = videoFrame.metadata, let timestampe = String(data: metadata, encoding: .utf8) else {
+            print("Receiving video frame without metadata attached")
+            return
+        }
+        
+        DispatchQueue.main.async {
+            self.metadataLabel.text = timestampe
+            print("Receiving video frame metadata", timestampe)
+        }
+    }
+}
+
+extension ViewController: FrameCapturerMetadataDelegate {
+    func finishPreparingFrame(_ videoFrame: OTVideoFrame?) {
+        guard let videoFrame = videoFrame else {
+            return
+        }
+        setTimestampToVideoFrame(videoFrame)
+    }
+    
+    fileprivate func setTimestampToVideoFrame(_ videoFrame: OTVideoFrame?) {
+        guard let videoFrame = videoFrame else {
+            return
+        }
+        
+        let dateFormatter = DateFormatter()
```

Comment:
you could have the formatter as an instance variable instead of recreating it here 30 times per second :)

Reply:
good catch! and this is indicated as a very expensive operation by Apple.